Add a mechanism for communicating the size of a Group ID gap (#922)
moq-wg/moq-transport: draft-ietf-moq-transport.md

~~~

# Extension Headers

> The following Object Extension Headers are defined in MoQT.

Author: The motivation for using Extension Headers is largely about keeping the SUBGROUP_HEADER cleaner, but could also be spelled that way.

> be updated if MOQT is also going to start defining extensions?

Let's gather more feedback here, but if we define some "core extensions" that are in moqt we will update that.

> The following Object Extension Headers are defined in MoQT.

Reviewer: We don't expect gaps to be on most Objects, so a header seems appropriate

> The following Object Extension Headers are defined in MoQT.

Reviewer 2: I am fine with extension given its applies to specific flows or conditions. Having said that, I am ok with anything that works for streams as well as datagrams and not bloat the latter, where not needed.

> ## Group ID Gap Size

Reviewer: Or maybe "Prior Group Gap"?  Size is somewhat implied.

Group ID Gap Size (Extension Header Type 0x40) is a variable length integer
containing the number of Groups prior to the current Group that do not and will
never exist. For example, if the Original Publisher published an Object in Group
7 and knows it will never publish any Objects in Group 8 or Group 9, it can
include Group ID Gap Size = 2 in any Object in Group 10.  A Group that contains
more than one Object with different values for Group ID Gap Size is considered

> malformed.  If this extension is present in an Object in Group 0, the value

Reviewer: As discussed, please also discuss the case when the Gap value is larger than the Group ID

> malformed.  If this extensions is present in an Object in Group 0, the value

Reviewer 3: A solution to this may be as follows:
`If there are more than one Object with different values for Group ID Gap size, the Object with the lowest ID is used to set Group ID Gap Size.`

> malformed.  If this extensions is present in an Object in Group 0, the value

Author: > How should the receiver react to a malformed group? 

I'll file a separate issue.  There are lots of ways a track or group can be malformed today.

> include Group ID Gap Size = 2 in any Object in Group 10.  A Group that contains
> more than one Object with different values for Group ID Gap Size is considered
> malformed.  If this extension is present in an Object in Group 0, the value
> MUST be ignored.

Author: ```suggestion
include Group ID Gap Size = 2 in any Object in Group 10.  A Group that contains
more than one Object with different values for Group ID Gap Size, has a
Group ID Gap Size larger than or equal to the Group ID, or has an Object present
in a Group ID within a gap is considered malformed.
```

> in a Group ID within a gap is considered malformed.

Reviewer 2: I am not sure I fully parse the last sentence. May be an example might help here.

Author: > has an Object present in a Group ID within a gap is considered malformed.

This part?  What I mean is: I receive Object {10,0} with Group ID Gap Size=1, then I receive Object {9,0}, but the publisher told me Group ID 9 does not exist.

Reviewer: It's quite a long sentence, so how about splitting it up? I think I'd split up the "Has an Object present ..." clause because it's a fairly different case from the first two.

Also, can the gap not be equal to the Group ID?  ie: Can Group 2 not have a Gap Size of 1?

Author: I think you're right, equal is ok but it can't be greater than the group ID:

Group 1 can have a Gap size 1 (eg, Group 0 does not exist)

> This extension is not mandatory, as publishers might not know the Gap Size. This

Author: ```suggestion
This extension is optional, as publishers might not know the Gap Size. This
```

extension can be added by the Original Publisher, but cannot be added by relays.

> This extension cannot be removed.

Author: ```suggestion
This extension cannot be modified or removed.
```

> extension can be added by the Original Publisher, but cannot be added by relays.
> This extension cannot be modified or removed.

Author: ```suggestion
extension can be added by the Original Publisher, but MUST NOT be added by
relays. This extension MUST NOT be modified or removed.
```

~~~

# Extension Headers

The following Object Extension Headers are defined in MoQT.

## Prior Group ID Gap

Prior Group ID Gap (Extension Header Type 0x40) is a variable length integer
containing the number of Groups prior to the current Group that do not and will
never exist. For example, if the Original Publisher published an Object in Group
7 and knows it will never publish any Objects in Group 8 or Group 9, it can

> include Prior Group ID Gap = 2 in any Object in Group 10.  A track with a Group

Reviewer 2: ```suggestion
include Prior Group ID Gap = 2 in any number of Objects in Group 10, as it sees fit.  A track with a Group
```

> include Prior Group ID Gap = 2 in any number of Objects in Group 10, as it sees fit.  A track with a Group

Author: That's true.  It's also true for other object Statuses that might be more important like End of Group/End of Track.

> that contains more than one Object with different values for Prior Group ID Gap,

Reviewer 2: ```suggestion
that contains more than one Object with different values for Prior Group ID Gap or
```

has a Prior Group ID Gap larger than the Group ID is considered malformed.  If an
endpoint receives an Object with a Group ID within a previously communicated gap

> is also treats the track as malformed.

Reviewer 2: ```suggestion
treats the track as malformed.
```

> is also treats the track as malformed.

Author: ```suggestion
it also treats the track as malformed.
```

> This extension is optional, as publishers might not know the prior gap gize. If

Reviewer: ```suggestion
This extension is optional, as publishers might not know the prior gap gize or there might not be a gap. If
```

Prior Group ID Gap is not present, the receiver cannot infer any information
about the existence of prior groups (see {{group-ordering}}).

This extension can be added by the Original Publisher, but MUST NOT be added by

> relays. This extension MUST NOT be modified or removed.

Reviewer 2: I get the intention. There is no easy to enforce this unless there is a way to verify the integrity between original publisher and the end subscriber.

Author: That will be true of every extension defined in MoQT, right?
